src/Ldap/Client.hs

```diff
@@ -139,51 +147,58 @@ data Disconnect = Disconnect Type.ResultCode Dn Text
 
 instance Exception Disconnect
 
+connect :: Host -> PortNumber -> IO (Ldap b)
+connect host port = do
+  context <- Conn.initConnectionContext
+  conn <- Conn.connectTo context (params host port)
+  l <- newLdap conn
+  return l
+
+close :: Ldap b -> IO ()
+close l@(Ldap client v conn threads) = do
+  unbindAsync l
+  forM threads Async.cancel
+  Conn.connectionClose conn
+
 -- | The entrypoint into LDAP.
 --
 -- It catches all LDAP-related exceptions.
-with :: Host -> PortNumber -> (Ldap -> IO a) -> IO (Either LdapError a)
+with :: Host -> PortNumber -> ((Ldap a) -> IO a) -> IO (Either LdapError a)
```

Comment:
It would be reasonable to reuse `connect`/`close` here to avoid code duplication.